src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java

```diff
@@ -18,53 +18,67 @@
 package org.mqttbee.mqtt.message.publish;
 
 import org.mqttbee.annotations.NotNull;
+import org.mqttbee.api.mqtt.datatypes.MqttTopic;
 import org.mqttbee.api.mqtt.mqtt5.message.publish.TopicAliasUsage;
-import org.mqttbee.mqtt.datatypes.MqttTopicImpl;
 
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Random;
 
+import static org.mqttbee.mqtt.message.publish.MqttPublishWrapper.DEFAULT_NO_TOPIC_ALIAS;
+
 /**
  * @author Silvio Giebl
  */
 public class MqttTopicAliasMapping {
 
     private static final Random random = new Random();
 
-    private final int size;
+    private final int topicAliasMaximum;
     private final HashMap<String, Integer> hashMap;
     private int nextTopicAlias;
 
-    public MqttTopicAliasMapping(final int size) {
-        this.size = size;
-        hashMap = new HashMap<>(size);
+    public MqttTopicAliasMapping(final int topicAliasMaximum) {
+        this.topicAliasMaximum = topicAliasMaximum;
+        hashMap = new HashMap<>(topicAliasMaximum);
         nextTopicAlias = 1;
     }
 
-    public int set(@NotNull final MqttTopicImpl topic, @NotNull final TopicAliasUsage topicAliasUsage) {
-        int topicAlias = MqttPublishWrapper.DEFAULT_NO_TOPIC_ALIAS;
-        if (topicAliasUsage != TopicAliasUsage.MUST_NOT) {
-            if (nextTopicAlias == size) {
-                if (topicAliasUsage == TopicAliasUsage.MAY_OVERWRITE) {
-                    topicAlias = random.nextInt(size) + 1;
-                    hashMap.put(topic.toString(), topicAlias);
-                }
-            } else {
-                topicAlias = nextTopicAlias;
-                hashMap.put(topic.toString(), topicAlias);
-                nextTopicAlias++;
-            }
+    public int set(@NotNull final MqttTopic topic, @NotNull final TopicAliasUsage topicAliasUsage) {
+        if (topicAliasUsage == TopicAliasUsage.MUST_NOT) {
+            return DEFAULT_NO_TOPIC_ALIAS;
+        }
+
+        // found existing
+        int topicAlias = get(topic);
+        if (topicAlias != DEFAULT_NO_TOPIC_ALIAS) {
+            return topicAlias;
         }
+
+        if (nextTopicAlias > topicAliasMaximum &&
+                topicAliasUsage == TopicAliasUsage.MAY_OVERWRITE) {
+            // used up all allowed topic aliases. Can only set a new one
+            // if allowed to overwrite an existing one.
+            topicAlias = random.nextInt(topicAliasMaximum) + 1;
+            // first remove the old mapping
+            hashMap.values().remove(topicAlias);
+            hashMap.put(topic.toString(), topicAlias);
+        } else {
```

Comment:
This is not correct. In the following case we are assigning a topic alias which is greater than the topicAliasMaximum:
nextTopicAlias > topicAliasMaximum and topicAliasUsage != TopicAliasUsage.MAY_OVERWRITE